hkdf.go

```
@@ -119,6 +122,18 @@ func ExtractHKDF(h func() hash.Hash, secret, salt []byte) ([]byte, error) {
 		return nil, err
 	}
 
+	// If calling code specifies nil salt, replace it with a buffer of hashLen
+	// zeros, as specified in RFC 5896 and as OpenSSL EVP_KDF-HKDF documentation
+	// instructs. Take a slice of a preallocated buffer to avoid allocating new
+	// buffer per call.
+	if len(salt) == 0 {
+		hlen := h().Size()
+		if hlen > len(hkdfAllZerosSalt) {
+			return nil, errors.New("nil salt is not supported for the specified hash algorithm")
```

In theory, this could reduce compatibility with existing providers, right? If the provider supports empty salt, but golang-fips/openssl doesn't have enough buffer space, then returning this error means the call will fail when it could have succeeded if we were to fall back to passing an empty salt.

However, if we let the call continue rather than returning an error and the call *does* fail, then the user ends up with an unclear error message.

Alternative: we could handle this case safely by allocating memory if the buffer isn't large enough. It might be slower for a provider that could have handled an empty salt, but at least it isn't a crash. (We can then add tests that make sure that for all known hashes, it doesn't allocate, for futureproofness.)

I don't know if this is a practical concern in context.

---

I would just make the preallocated buffer larger, say 512 bytes.

---

The largest hash size we support is 64, so I would stick to that. An option that takes the best of the two suggestions is doung something like this:

```go
if len(salt) == 0 {
  salt = make([]byte, h.Size(), maxHashSize)
}
```

Note that `maxHashSize` is already defined in `hash.go`. That construct doesn't allocate if `h.Size()` is smaller than `maxHashSize`, and allocates if it is (which shouldn't happen). This way we support all cases and cap remove the error return.